test/common/filter/config_discovery_impl_test.cc

```diff
@@ -127,311 +119,364 @@ config_source: { ads: {} }
+  }
+  const std::string getFilterType() const override { return "http"; }
+  const std::string getConfigReloadCounter() const override {
+    return "extension_config_discovery.http_filter.foo.config_reload";
+  }
+  const std::string getConfigFailCounter() const override {
+    return "extension_config_discovery.http_filter.foo.config_fail";
+  }
+};
+
+// TCP listener filter test
+class TcpListenerFilterConfigDiscoveryImplTest
+    : public FilterConfigDiscoveryImplTest<
+          Network::ListenerFilterFactoryCb, Server::Configuration::ListenerFactoryContext,
+          TcpListenerFilterConfigProviderManagerImpl, Envoy::ProtobufWkt::Struct> {
+public:
+  const std::string getTypeUrl() const override { return "google.protobuf.Struct"; }
+  const std::string getFilterType() const override { return "listener"; }
+  const std::string getConfigReloadCounter() const override {
+    return "extension_config_discovery.tcp_listener_filter.foo.config_reload";
+  }
+  const std::string getConfigFailCounter() const override {
+    return "extension_config_discovery.tcp_listener_filter.foo.config_fail";
+  }
+};
+
+// UDP listener filter test
+class UdpListenerFilterConfigDiscoveryImplTest
+    : public FilterConfigDiscoveryImplTest<
+          Network::UdpListenerFilterFactoryCb, Server::Configuration::ListenerFactoryContext,
+          UdpListenerFilterConfigProviderManagerImpl,
+          test::integration::filters::TestUdpListenerFilterConfig> {
+public:
+  const std::string getTypeUrl() const override {
+    return "test.integration.filters.TestUdpListenerFilterConfig";
+  }
+  const std::string getFilterType() const override { return "listener"; }
+  const std::string getConfigReloadCounter() const override {
+    return "extension_config_discovery.udp_listener_filter.foo.config_reload";
+  }
+  const std::string getConfigFailCounter() const override {
+    return "extension_config_discovery.udp_listener_filter.foo.config_fail";
+  }
+};
+
+/***************************************************************************************
+ *                  Parameterized test for                                             *
+ *     HTTP filter,  TCP listener filter And UDP listener filter                       *
+ *                                                                                     *
+ ***************************************************************************************/
+template <typename FilterConfigDiscoveryTestType>
+class FilterConfigDiscoveryImplTestParameter : public testing::Test {
+public:
+  FilterConfigDiscoveryTestType config_discovery_test_;
+};
+
+// The test filter types.
+using FilterConfigDiscoveryTestTypes =
+    ::testing::Types<HttpFilterConfigDiscoveryImplTest, TcpListenerFilterConfigDiscoveryImplTest,
+                     UdpListenerFilterConfigDiscoveryImplTest>;
+
+TYPED_TEST_SUITE(FilterConfigDiscoveryImplTestParameter, FilterConfigDiscoveryTestTypes);
```

Comment:
Interesting... I actually meant that you can use a parameterized test similar to:
https://github.com/envoyproxy/envoy/blob/ba362a22b43a41379b156563f2f83402762081bf/test/integration/multiplexed_integration_test.cc#L46-L50

but this also looks ok to me.

Reply:
Thanks for the comments.

In my case, the variables is type of the data, not data itself.  For typed parameterized test, GTEST is using below machinery, which is what's used in this PR:

https://google.github.io/googletest/advanced.html#typed-tests